[JSC] Avoid calling setIsFunction() for function name bindings

shvaikalesh opened the pull request:
#### f82533b7e079ad511df75a66c942f60ec73c145a
<pre>
[JSC] Avoid calling setIsFunction() for function name bindings
<a href="https://bugs.webkit.org/show_bug.cgi?id=263097">https://bugs.webkit.org/show_bug.cgi?id=263097</a>
&lt;rdar://problem/116891657&gt;

Reviewed by Ross Kirsling.

Our implementations of GlobalDeclarationInstantiation / EvalDeclarationInstantiation
(ProgramExecutable::initializeGlobalProperties() / Interpreter::executeEval() respectively)
rely on IsFunction to detect bindings created via FunctionDeclaration &amp; friends.

While calling setIsFunction() on a function name binding doesn&apos;t seem to affect any logic expect for
duplicate parameter SyntaxError, it&apos;s semantically more pure to reserve IsFunction bit only for
FunctionDeclaration &amp; friends.

Since at the time of declareParameter() being called, only the function name binding and other
parameters can be declared, this change tweaks the SyntaxError condition to utilize isParameter()
instead, ensuring that it&apos;s checked before calling setIsParameter(), and refactors `isValidStrictMode`
to rely on that condition as well.

No new tests, no behavior change.

* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::declareCallee):
(JSC::Scope::declareParameter):

Canonical link: <a href="https://commits.webkit.org/269379@main">https://commits.webkit.org/269379@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96efc7a2f20f26e5c85ad9309f4ff1e886877baf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22127 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22335 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24019 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20484 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26586 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22652 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21517 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22357 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19194 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24869 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19118 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20050 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26307 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19236 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20137 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20275 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24166 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21497 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20710 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17630 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25542 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20064 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6018 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24272 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26818 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2803 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20658 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5850 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rkirsling reviewed (approved):
Seems sensible to me, as long as tests are green.

hyjorc1 reviewed (approved):
Nice! LGTM!

Commit:
[JSC] Avoid calling setIsFunction() for function name bindings
https://bugs.webkit.org/show_bug.cgi?id=263097
<rdar://problem/116891657>

Reviewed by Ross Kirsling.

Our implementations of GlobalDeclarationInstantiation / EvalDeclarationInstantiation
(ProgramExecutable::initializeGlobalProperties() / Interpreter::executeEval() respectively)
rely on IsFunction to detect bindings created via FunctionDeclaration & friends.

While calling setIsFunction() on a function name binding doesn't seem to affect any logic expect for
duplicate parameter SyntaxError, it's semantically more pure to reserve IsFunction bit only for
FunctionDeclaration & friends.

Since at the time of declareParameter() being called, only the function name binding and other
parameters can be declared, this change tweaks the SyntaxError condition to utilize isParameter()
instead, ensuring that it's checked before calling setIsParameter(), and refactors `isValidStrictMode`
to rely on that condition as well.

No new tests, no behavior change.

* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::declareCallee):
(JSC::Scope::declareParameter):

Canonical link: https://commits.webkit.org/269379@main
